pep-0677.rst

```diff
@@ -839,6 +860,90 @@ existing function header syntax. Moreover, it is visually similar to
 lambdas, which bind names with no parentheses: ``lambda x, y: x ==
 y``.
 
+Requiring Outer Parentheses
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A concern with the current proposal is readability, particularly
+when callable types are used in return type position which leads to
+multiple top-level ``->`` tokens, for example::
+
+    def f() -> (int) -> (int) -> int
+        return lambda x: lambda y: x + y
+
+We considered a few ideas to prevent this by changing rules about
+parentheses. One was to move the parentheses to the outside, so
+that a two-argument boolean function is written ``(int, str -> bool)``.
+With this change, the example above becomes::
+
+    def f() -> (int -> (int -> int)):
+        return lambda x: lambda y: x + y
+
+This makes the nesting of many examples that are difficult to
+follow clear, but we rejected it because
+
+- Currently in Python commas bind very loosely, which means it might be common
+  to misread ``(int, str -> bool)`` as a tuple whose first element is an int,
+  rather than a two-parameter callable type.
+- It is not very similar to function header syntax, and one of our goals was
+  familiar syntax inspired by function headers.
+- This syntax may be more readable for deaply nested callables like the one
+  above, but deep nesting is not very common. Encouraging extra parentheses
+  around callable types in return position via a style guide would have most of
+  the readability benefit without the downsides.
+
+We also considered requiring parentheses on both the parameter list and the
+outside, e.g. ``((int, str) -> bool)``. With this change, the example above
+becomes::
+
+    def f() -> ((int) -> ((int) -> int)):
+        return lambda x: lambda y: x + y
+
+We rejected this change because:
+
+- The outer parentheses only help readability in some cases, mostly when a
+  callable type is used in return position. In many other cases they hurt
+  readability rather than helping.
+- We agree that it might make sense to encourage outer parentheses in several
+  cases, particularly callable types in function return annotations. But we
+  believe it is more appropriate to encourage this in style guides, linters,
+  and autoformatters than to bake it into the parser and throw syntax errors.
```

Comment:
Also, this entire example looks unPythonic, and even if such a function would make sense, I'd probably use a type alias to make the return type less look like line noise.

Reply:
Simplified the example a bit and added a bullet about type aliasing as a solution.

Although I have some skepticism about type aliases as a general answer to complex types. Based on the production typed-python code I see I, developers seem hesitant to use them (at least at Meta).

I can understand why. Most code forms a graph where you constantly have to jump-to-source to understand it. One nice thing about types is that they normally speak for themselves. Once we use type aliases, that becomes less true - readers might have to jump hundreds of lines to the top of a module to see what a type actually represents.

This isn't necessarily a bad thing since type aliases provide abstraction and one kind of readability, but there is definitely a cost and it seems to me that many developers don't want to pay it.

Reply:
Yeah, it depends on lots of things. But I don’t want to feel that we have to do something just because Facebook coders prefer it. There’s a whole community here, and some people are already concerned that Python is more and more being controlled by enterprise developers.